web3/utils/datastructures.py

```diff
@@ -0,0 +1,70 @@
+
+from collections import (
+        Mapping,
+        MutableMapping,
+        Hashable,
+        )
+
+# Hashable must be immutable:
+# "the implementation of hashable collections requires that a key's hash value is immutable"
+# https://docs.python.org/3/reference/datamodel.html#object.__hash__
+
+
+class ReadableAttributeDict(Mapping):
+    'the read attributes for the AttributeDict types'
+
+    def __init__(self, dictionary, *args, **kwargs):
+        self.__dict__ = dict(dictionary)
+        self.__dict__.update(dict(*args, **kwargs))
+
+    def __getitem__(self, key):
+        return self.__dict__[key]
+
+    def __iter__(self):
+        return iter(self.__dict__)
+
+    def __len__(self):
+        return len(self.__dict__)
+
+    def __str__(self):
+        return repr(self)
+
+    def __repr__(self):
+        dict_lines = ['\n\t' + repr(k) + ': ' + repr(self[k]) + ',' for k in sorted(self.__dict__)]
```

Comment:
I'm 👍 on pretty docs examples but those can be massaged by hand or generated using the `pprint` module.  I'm even inclined to not have it be wrapped in `AttributeDict(....)` and to just pass it off a normal dictionary when repr'd in the console for simplicity sake.  I'm -0 on this which should be interpeted as "I don't agree but not strongly enough that we need to reach consensus.".

Your call 😛